tests/ut/plugin_demo/customIdx/customIdx.cc

```diff
@@ -0,0 +1,617 @@
+namespace calculator {
+template <typename VecT>
+inline float
+vector_norm(const VecT* x, int32_t dim) {
+    float x_norm_square = 0.0;
+    for (int32_t i = 0; i < dim; i++) {
+        x_norm_square += float(x[i]) * float(x[i]);
+    }
+    return sqrt(x_norm_square);
+}
+
+template <typename VecT>
+inline float
+l2_distance(const VecT* x, const VecT* y, int32_t dim) {
+    float ret = 0.0;
+    for (int32_t i = 0; i < dim; i++) {
+        float diff = float(x[i]) - float(y[i]);
+        ret += diff * diff;
+    }
+    return ret;
+}
+
+template <typename VecT>
+inline float
+ip_distance(const VecT* x, const VecT* y, int32_t dim) {
+    float ret = 0.0;
+    for (int32_t i = 0; i < dim; i++) {
+        float dot = float(x[i]) * float(y[i]);
+        ret += dot;
+    }
+    return ret;
+}
+
+template <typename VecT>
+inline float
+cosine_distance(const VecT* x, const VecT* y, int32_t dim) {
+    float x_norm = vector_norm(x, dim);
+    float y_norm = vector_norm(y, dim);
+
+    float x_y_inner = ip_distance(x, y, dim);
+    if (std::abs(x_norm) > std::numeric_limits<float>::epsilon() &&
+        std::abs(y_norm) > std::numeric_limits<float>::epsilon()) {
+        return x_y_inner / (x_norm * y_norm);
+    }
+    return 0.0;
+}
+
+template <typename VecT>
+auto
+get_distance_calculator(knowhere::MetricType metricType) {
+    if (metricType == knowhere::metric::L2) {
+        return l2_distance<VecT>;
+    } else if (metricType == knowhere::metric::COSINE) {
+        return cosine_distance<VecT>;
+    } else if (metricType == knowhere::metric::IP) {
+        return ip_distance<VecT>;
+    } else {
+        LOG_KNOWHERE_ERROR_ << "Only support {L2, IP, COSINE} metric. " << metricType << " does not support";
+        throw knowhere::KnowhereException("Unsupported Metric");
+    }
+}
+
+bool
+l2_distance_range_checker(const float dist, const float radius, const float range_filter) {
+    return range_filter <= dist && dist < radius;
+}
+
+bool
+ip_distance_range_checker(const float dist, const float radius, const float range_filter) {
+    return radius < dist && dist <= range_filter;
+}
+
+bool
+cosine_distance_range_checker(const float dist, const float radius, const float range_filter) {
+    return radius < dist && dist <= range_filter;
+}
+
+auto
+get_distance_range_checker(knowhere::MetricType metricType) {
+    if (metricType == knowhere::metric::L2) {
+        return l2_distance_range_checker;
+    } else if (metricType == knowhere::metric::COSINE) {
+        return cosine_distance_range_checker;
+    } else if (metricType == knowhere::metric::IP) {
+        return ip_distance_range_checker;
+    } else {
+        LOG_KNOWHERE_ERROR_ << "Only support {L2, IP, COSINE} metric. " << metricType << " does not support";
+        throw knowhere::KnowhereException("Unsupported Metric");
+    }
+}
+
+}  // namespace calculator
+
+template <typename VecT>
+class CustomIdx : public knowhere::IndexNode {
```

Comment:
Please add a note on how the index is used. Basically, a user sees a constructor, `Train`, `Add`, `Search`. A typical user would think "let me construct, Add points, then Train, and then I will Search", right? 
So, the sequence needs to be mentioned and the reasoning behind it as well. Something like that `Train()` is like building a house, and once it is built, we may `Add()` people in it and then `Search()`.
Also, can 'Add()' be called multiple times, such as adding people one by one into the house, something that Faiss allows, or not?

Reply:
@liliu-z is about to reshap the interface in the next few days. We will just remain `Build` for immutable index, while growing index will use `Build` and `Add` for first initialization and subsequent insertions.